File: Trees/98. 验证二叉搜索树(含無限條件)/98. 验证二叉搜索树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        long long inorder = (long long)INT_MIN - 1;
        stack<TreeNode*> stk;

        while(root != nullptr || !stk.empty()){
            while(root != nullptr){
                stk.push(root);
                root = root->left;
            }

            TreeNode* node = stk.top();
            stk.pop();

            if(node->val <= inorder) return false;

            inorder = node->val;
            root = node->right;
        }
        return true;
    }
};

File: Trees/98. 验证二叉搜索树(含無限條件)/98. 验证二叉搜索树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        temp = float('-inf')
        stk = []

        while root or stk:
            while root:
                stk.append(root)
                root = root.left
            
            node = stk.pop()

            if node.val <= temp:
                return False
            temp = node.val

            root = node.right
        
        return True